Mathlib/CategoryTheory/Sites/Coherent.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+
+import Mathlib.CategoryTheory.Sites.Coverage
+import Mathlib.CategoryTheory.Sites.EffectiveEpimorphic
+
+/-!
+
+# The Coherent Grothendieck Topology
+
+This file defines the coherent Grothendieck topology (and coverage) on a category `C`.
+The category `C` must satisfy a `Precoherent C` condition, which is essentially the minimal
+requirement for the coherent coverage to exist.
+
+In `isSheaf_coherent`, we characterize the sheaf condition for presheaves of types for the
+coherent Grothendieck topology in terms of finite effective epimorphic families.
+
+-/
+
+namespace CategoryTheory
+
+open Limits
+
+variable (C : Type _) [Category C]
+
+/--
+The condition `Precoherent C` is essentially the minimal condition required to define the
+coherent coverage on `C`.
+-/
+class Precoherent : Prop where
+  /--
+  Given an effective epi family `π₁` over `B₁` and a morphism `f : B₂ ⟶ B₁`, there exists
+  an effective epi family `π₂` over `B₂`, such that `π₂` factors through `π₁`.
+  -/
+  pullback {B₁ B₂ : C} (f : B₂ ⟶ B₁) :
+    ∀ (α : Type) [Fintype α] (X₁ : α → C) (π₁ : (a : α) → (X₁ a ⟶ B₁)),
+      EffectiveEpiFamily X₁ π₁ →
+    ∃ (β : Type) (_ : Fintype β) (X₂ : β → C) (π₂ : (b : β) → (X₂ b ⟶ B₂)),
+      EffectiveEpiFamily X₂ π₂ ∧
+      ∃ (i : β → α) (ι : (b :  β) → (X₂ b ⟶ X₁ (i b))),
+      ∀ (b : β), ι b ≫ π₁ _ = π₂ _ ≫ f
+/--
+The coherent coverage on a precoherent category `C`.
+-/
+def CoherentCoverage [Precoherent C] : Coverage C where
```

Comment:
Why not `CoherentCoverage`? I understand it is long, but you are not in a `Coherent` namespace, right?

Reply:
This is now done.